### DISCUSSION TITLE
simple DOS EXE works but DOS COM crashes

### DISCUSSION THREAD

**Discussion by LowLevelMahn** [open-watcom/open-watcom-v2#821] [Q&A] [answered]
using: https://github.com/open-watcom/open-watcom-v2/releases/tag/Current-build, Last updated 2022-02-20 01:50:29

hello.c

```
#include <stdio.h>
void main()
{
  printf("hello\n");
}
```

build.bat

```
set WATCOM=%tools_dir%\open-watcom-2_0-c-win-x64
set WATCOM_BIN=%watcom%\binnt64
set INCLUDE=%watcom%\h
set PATH=%WATCOM_BIN%;%PATH%
set wcc_exe=%watcom_bin%\wcc.exe
set wlink_exe=%watcom_bin%\wlink.exe

%wcc_exe% hello.c
%wlink_exe% name hello.com system dos com file { hello.obj }
```

gives 

```
Open Watcom C x86 16-bit Optimizing Compiler
Version 2.0 beta Feb 20 2022 00:59:21 (64-bit)
Copyright (c) 2002-2022 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1984-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
hello.c: 8 lines, included 813, 0 warnings, 0 errors
Code size: 17
Open Watcom Linker Version 2.0 beta Feb 20 2022 00:55:27 (64-bit)
Copyright (c) 2002-2022 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
searching libraries
creating a DOS .COM executable
Warning! W1019: file clibs.lib(cstart): segment relocation at 0000:00a9
Warning! W1019: file clibs.lib(cstart): segment relocation at 0000:0135
Warning! W1019: file clibs.lib(cstart): segment relocation at 0000:021a
Warning! W1019: file clibs.lib(cstart): segment relocation at 0000:02fd
Warning! W1087: stack segment ignored in .COM file

```

and the COM file crashes 

works when creating a exe with `system dos`

**Comment by jmalak** [comment]
Sorry, it is not a problem with OW.
COM format is special DOS format with single segment only.
See documentation [Creating a Tiny Memory Model Application](https://open-watcom.github.io/open-watcom-v2-wikidocs/cguide.html#Creating_a_Tiny_Memory_Model_Application) .

**Comment by LowLevelMahn** [comment]
thanks

`wlink name hello.com system dos com file { %WATCOM%\lib286\dos\cstart_t.obj hello.obj }`

creates a non-crashing hello.com

question: why does `system dos com` not implicitly add cstart_t.obj as first obj file?

**Comment by jmalak** [comment]
Read documentation carefully there is option **`-ms`** for compilation sources and linker command is also different, uses **`system com`**.
The right commands in your build.bat (which do all what you expect) are

%wcc_exe% hello.c **`-ms`**
%wlink_exe% name hello.com system **`com`** file { hello.obj }

against yours

%wcc_exe% hello.c **` `**
%wlink_exe% name hello.com system **`dos com`** file { hello.obj }

For cross-compilation you should add **`-bt=dos`** option for compilers too.
Your compiler command should looks like
%wcc_exe% hello.c **`-bt=dos -ms`**

If you don't specify **`-bt=dos`** option then implicit value is taken, by example if you compile on OS/2 then implicit equivalent of **`-bt=os2`** is used not **`-bt=dos`**. It is good practice to specify **`-bt=...`** option on command line then command is cross-compile save.

**Comment by LowLevelMahn** [comment]
Thank your clearing that up - there are so many options and combination that sometimes/or partially work - trial & error isn't really a good strategie here :)

**Comment by jmalak** [comment]
Take into account that OW is cross-compiler by nature against Microsoft compilers or GCC which works by default for single target.
You need not to use too many option but you should always specify 3 things explicitly, even if it is defined implicitly for host system:
- compiler target **`-bt=...`** option  (implicit is host system)
- compiler memory model **`-m...`** option (implicit is **`-ml`** for 16-bit, **`-mf`** or **`-ms`** for 32-bit, depend on target)
- linker target executable format **`system ....`** directive (implicit is host system for 32-bit and **`system dos`** is for 16-bit if host is 32-bit only, by example Linux host)

**Comment by LowLevelMahn** [comment, marked as answer]
thanks for the info
